framework/include/base/ObjectStorage.h

```diff
@@ -0,0 +1,106 @@
+/****************************************************************/
+/*               DO NOT MODIFY THIS HEADER                      */
+/* MOOSE - Multiphysics Object Oriented Simulation Environment  */
+/*                                                              */
+/*           (c) 2010 Battelle Energy Alliance, LLC             */
+/*                   ALL RIGHTS RESERVED                        */
+/*                                                              */
+/*          Prepared by Battelle Energy Alliance, LLC           */
+/*            Under Contract No. DE-AC07-05ID14517              */
+/*            With the U. S. Department of Energy               */
+/*                                                              */
+/*            See COPYRIGHT for full restrictions               */
+/****************************************************************/
+
+#ifndef OBJECTSTORAGE_H
+#define OBJECTSTORAGE_H
+
+// MOOSE includes
+#include "MooseTypes.h"
+
+/**
+ * A general storage container for objects capable of storing threaded copies.
+ */
+template<typename ObjectType>
+class ObjectStorage
```

Comment:
I'm having a difficult time seeing the need for ObjectStorage and MooseObjectWarehouse